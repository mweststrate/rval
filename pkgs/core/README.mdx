---
name: Core api
order: 1
menu: API
route: /api/core
---

# @r-val/core

`@r-val/core` package contains all the core functionality of RVal. 
The most important concepts are explain on the [getting started](#/introduction/the-basics) page.

## `val`

**Signature** 

`val(initialValue, preProcessor?) -> reactive value`

**Description**

`val` creates a reactive value that can be used to store state.
The `initialValue` froms the initial state of the reative value.

`val` returns a function that takes zero or one arguments:
* If the function is invoked without arguments, the current value is returned
* If the function is invoked with one argument, and that value is not a function, the argument becomes the new state of the reactive value. (If you want to store a function itself as value, use `val(() => fn)` instead of `val(fn)`)
* If the function is invoked with a function as argument, that function is invoked and should produce the new state, based on the current state. Built-in update functions can be found in the [`@r-val/updaters` package](#/api/updaters).

```javascript
const temperature = val(37)
console.log(temperature()) // prints 37

temperature(38)
console.log(temperature()) // prints 38

temperature(v => v + 1)
console.log(temperature()) // prints 39
```

For the `preProcessor` argument, see [#/advanced/preprocessors]. `preProcessor` can both be a single pre-processor, or an array of pre-processors.

## `drv`

**Signature**

`drv(derivation, setter?) -> reactive value`

**Description**

`drv` creates a derived, reactive value. It accepts a single function (which should not have side effects),
and it will automatically subscribe to all reactive values that were accessed when running that function.
The output of the function is cached until one observed reactive values changes.

Generally, it is recommended to use `drv` abunduntaly; every `drv` is a potential caching point, making change propagation trough your system easier.
It will also help you reduce the amount of _state_ you have in your application, and under almost all circumstances it is better
to have a piece of derived information, over some redundant information that is kept up to date through other means (such as `sub`).

The optional `setter` argument can be used to make the derived value writable. 
This is just syntactic convenience, as the `drv` itself does not hold state, but it can be used as 'write-through' mechanism.
Setters are automatically wrapped in an `act`.

```javascript
const priceDollar = val(10)
const euroDollarRate = val(1.1487)

const priceEuro = drv(
    // derive the price in euro
    () => priceDollar() / euroDollarRate(),
    // setter: makes priceEuro writeable
    (newPriceEur) => {
        priceDollar(newPriceEur * euroDollarRate()) 
    }
)

console.log(priceEuro()) // 8.71
priceEuro(10)
console.log(priceDollar()) // 11.49
```

By default, RVal assumes that a `drv` that is not observed itself by anyone (by a `drv`, `sub` or `effect) is not relevant anymore,
and it will automatically suspend the computation, meaning that it will stop observing it's inputs on the next JavaScript tick.
This is done to make sure no memory leaks are introduced. If you want to forcefully cache the `drv` output under all circumstances, use `keepAlive` from `@r-val/utils`.

## `sub`

**Signature**

* `sub(source, listener, options) -> disposer` &mdash; the default way to create a subscription to an reactive `source`
* `sub(listener, options) -> (source) -> disposer` &mdash; create a curried subscription, 

**Subscription `options`:**

* `fireImmediately` (default `false`). `true`: the listener should be immediately invoked, or `false`: only when the first _change_ is observed in `source`.

**Listener arguments:**

* `value`
* `previousValue`

**Description**

`sub` subscribe to a reative value. 
It will trigger for each change in the reactive value, but only once per batch (see `act`).
`sub` is the primary means to create side effects that are triggered by reactive values,
such as logging, updating a rendering, or makign network requests.
(In the end, any system needs side effects, to be able to perform output).

`sub` returns a disposer function, that stops future executions of the `listener` once called.

```javascript
import { val, sub, drv } from "@r-val/core"

function createTodo(id, initialTitle) {
    const title = val(initialTitle)
    const done = val(false)
    const toJS = drv(() => ({
        id,
        title: title(),
        done: done()
    }))

    // set up a subscription to automatically save the 
    // todo when either title or done changes
    const stop = sub(toJS, data => {
        window.fetch("http://host/todos/" + id, { 
            method: "POST",
            body: JSON.stringify(data)
        }).catch(e => {
            console.error(e)
        }) 
    })

    return {
        title, done, toJS, stop
    }
}

const myTodo = createTodo(1, "get cofee")
myTodo.title("get coffee") // will be saved automatically
myTodo.stop() // stop future executions of the subscription
```

In cases, it might be tempting to update other reactive values in a subscription,
but in most cases this is an anti pattern and the value-to-be-updated could be expressed as derived value using `drv` instead.

The curried version of the `sub` can be useful to create utilities, for example a `logger` that prints the values of any state transition:

```javascript
const log = sub((current, previous) => {
    console.log(`${previous} -> ${current}`)
})

log(someReactiveValue)
log(someOtherReactiveValue)
```

N.B. this utility is actually built-in as `logChanges` in the `@r-val/utils` package.

## `act`

**signature**

`act(function) -> function` 

**discription**

`act` takes accepts a function, and returns a function with the same signature, that, when invoked, will batch the updates.
This is also called an _action_. 
This means, that if multiple mutations in that same function would normally affect multiple invocations of the same side-effect,
an function wrapped in `act` will trigger the side effects at most once. 

Although it is very well possible to create and invoke an action in one go (like: `act(() => { /* someStuff */ })()`),
it is recommended to not do this. Updates that are grouped together are a nice opportunity to extract a single function with limited functionality.

If functions that are wrapped by `act` invoke other functions wrapped by `act`, side effects will only trigger once the outer `act` completes.

That's all! Note that `act` only batches _synchronosly_ run updates. Passing an `async` function to `act` is in that regard mostly useless (only the code until the first `await` will be batched in that case).
Calling individual actions from `async` functions is no problem. 

## `effect`

**signature**

`effect(functionToTrack: () -> T, onInvalidate: (didChange() -> boolean, pull() -> T)): disposerFunction` 

**description**

`effect` is a core building block to organize side effects. 
For example `sub` and `@r-val/react`'s `rview` use `effect` under the hood.
`effect` tracks a function (that might produce a value), and notifies the `onInvalidate` callback when 
some of the dependencies of the tracked function _might_ have changed.

`effect` basically produces a state machine with the following states / contract (it initializes at step 7!):

1. The `functionToTrack` is executed, and the effect records which reactive values where used during the executions
2. If one or more of those dependencies might have changed, `onInvalidate` is called, with two parameters; `didChange()` and `pull()`
3. At any point in time, `didChange()` should be called. This will make sure that the dependencies are actually _evaluated_ (the effect might be depending on derived values that have not been evauated so far).
4. If `didChange()` returns `false`, nothing did actually changed, and `onInvalidate` might be invoked again if dependencies change in the future. (Go back to 2.)
5. If `didChange()` returns `true`, `pull()` should be called at any point in time in the future. `pull()` will perform step 1 again (tracking the function), register the dependencies, and returns any value it yields.
6. If the `disposer` returned from `effect` is called at any moment in time, the effect is immediately disposed, cleaning up the dependency tree.
7. Initially, the state machine is kicked of by calling `onInvalidate` once, starting basically at step 3. (Note that `didChange` will always yield `true` the first time)


Note that `onInvalidate` won't be called again, until `didChange()` is called and returned `false`, 
or `pull()` is called.

You probably won't use `effect` much in practice (using `sub` is more straightforward), 
but `effect` is a good building block to create utility functions that have very specific caching or timing behavior.

For example, an alternative to `sub(source, listener)` that debounces the listener and at max triggers its listener
every x milliseconds could be defined as:

```javascript
function sub(src, listener, delay) {
    let lastSeen = undefined

    return effect(
        () => src(),
        (didChange, pull) => {
            if (didChange()) {
                setTimeout(() => {
                    const v = pull()
                    if (v !== lastSeen)
                        listener(v, lastSeen)
                    lastSeen = v
                }, delay);
            }
        }
    )
}
```

## `isVal`

## `isDrv`


---

## `configure`

**Signature**

`configure(options)`

**Configuration options**

* `deepFreeze`: automatically deeply freezes any values arrays or plain objects that are stored in a reactive value. Best enabled in development enviroments. Default: `true`. Should be disabled in production enviroments for optimal performance.

Set's the the options for the current RVal instance (see also `rval` and `defaultInstance`).


## `rval`

Signature: 

* `rval()` &mdash; creates a _new_ RVal instance
* `rval(reactiveValue)` &mdash; returns the RVal instances that created the given reactive value.

Returns an rval instance, that exposes the APIs:

* `val`
* `drv`
* `sub`
* `act`
* `effect`
* `configure`

## `defaultInstance`

The default RVal instance used by all the top-level api's exposed by the module. 
For more info about what an RVal instance is, see `rval`.

## `$RVal`
