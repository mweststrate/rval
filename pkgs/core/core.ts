export type SubListener<T = any> = (value: T, prevValue?: T) => void

export type Thunk = () => void

export type Disposer = Thunk

// Exposed, this might be useful for third part reflection based (dev) tools
export const $RVal = typeof Symbol === "undefined" ? "$RVal" : Symbol.for('$RVal')

export interface Observable<T = unknown> {
  (): T
}

export interface Drv<T = unknown, S = T> extends Observable<T> {
  (newValue: T | S): void
}

export interface Val<T = unknown, S = T> extends Observable<T> {
  (updater: (current: T) => T | S): void
  (newValue: T | S): void
}

interface RValContext {
  config: RValConfig
  isUpdating: boolean
  pending: Thunk[],
  currentlyComputingStack: Set<ObservableAdministration>[]
  currentlyComputing: Set<ObservableAdministration>
  isRunningReactions: boolean
  runPendingObservers()
}

export interface RValConfig {
  autoFreeze: boolean
}

interface ObservableAdministration {
  addListener(observer: Thunk)
  removeListener(observer: Thunk)
  get(): any
}

export type PreProcessor<T = unknown, S = T> = (newValue: T | S, baseValue?: T, api?: RValInstance) => T

export interface RValInstance {
  val<T>(initial: T): Val<T, T>
  val<T, S=T>(initial: S, preProcessor?: PreProcessor<T, S> | (PreProcessor<T, any>[])): Val<T, S>
  drv<T, S=T>(derivation: () => T, setter?: (value: S) => void): Drv<T>
  sub<T>(
    listener: SubListener<T>,
    options?: SubscribeOptions
  ): (src: Observable<T>) => Disposer
  sub<T>(
    src: Observable<T>,
    listener: SubListener<T>,
    options?: SubscribeOptions
  ): Disposer
  effect<T>(fn: () => T, onInvalidate: (onChanged: () => boolean, pull: () => T) => void): Thunk
  act<T extends Function>(fn: T): T
  configure(config: Partial<RValConfig>): void
}

export interface SubscribeOptions {
  fireImmediately?: boolean
}

const NOT_TRACKING = 0
const STALE = 1
const UP_TO_DATE = 2

export function rval(base?: Val<any, any>): RValInstance {
  if (arguments.length) {
    if (!isVal(base) && !isDrv(base))
      throw new Error("Expected val as first argument to rval")
    return (base[$RVal] as any).api
  }

  const context: RValContext = {
    config: {
      autoFreeze: true 
    },
    isUpdating : false,
    pending: [],
    currentlyComputingStack: [],
    get currentlyComputing() {
      return this.currentlyComputingStack[this.currentlyComputingStack.length - 1]
    },
    isRunningReactions: false,
    runPendingObservers
  }

  function runAfterBatch(t: Thunk) {
    context.pending.push(t)
  }

  function val(initial, preProcessor: any = defaultPreProcessor): Val<any, any> {
    return new ObservableValue(context, api, initial, preProcessor).get as any
  }

  function drv<T>(derivation: () => T, setter: (value) => void): Drv<T> {
    return new Computed<T>(context, api, derivation, setter).get as any
  }

  function effect<T>(fn: () => T, onInvalidate: (didChange: () => boolean, pull: () => T) => void): Thunk {
    const computed = isDrv(fn) ? fn[$RVal] : new Computed(context, api, fn)
    let scheduled = true
    let disposed = false
    let lastSeen: any = {} // definitely not pointer equal to something else
    
    function didChange() {
      if (disposed) return false
      // the right hand of the OR is an easy in case the computed was evaluated early in a batch, but the effect didn't run, 
      // see test 'drv is not re-evaluating if triggered eagerly'
      const changed = computed.someDependencyHasChanged() || lastSeen !== computed.get()
      if (!changed) {
        scheduled = false // no pull is expected
      }
      return changed
    }
    function pull() {
      if (disposed) {
        throw new Error("[rval] pulling from already disposed effect")
      }
      scheduled = false
      return lastSeen = computed.get()
    }
    function onDirty () {
      if (scheduled || disposed) return
      scheduled = true
      runAfterBatch(() => onInvalidate(didChange, pull))
    }
    
    computed.addListener(onDirty)
    onInvalidate(didChange, pull)
    return _once(() => {
      disposed = true
      computed.removeListener(onDirty)
    })
  }

  function sub(
    src,
    listener?,
    options?
  ) {
    if (arguments.length === 1 || typeof arguments[1] !== "function") {
      // curried invocation
      return source => sub(source, src /* the listener actually */, listener /* the options actually */)
    }

    let lastSeen: any = undefined
    let firstRun = true
    return effect(src, (didChange, pull) => {
      if (didChange()) {
        const v = pull()
        if (firstRun ? (options && options.fireImmediately) : v !== lastSeen) 
          listener(v, lastSeen)
        lastSeen = v
        firstRun = false
      }
    })
  }

  function act<T extends Function>(fn: T): T {
    return function act(this: any) {
      if (context.isUpdating)
        return fn.apply(this, arguments)
      try {
        context.isUpdating = true
        return fn.apply(this, arguments)
      } finally {
        context.isUpdating = false
        runPendingObservers()
      }
    } as any
  }

  function runPendingObservers() {
    if (!context.isUpdating && !context.isRunningReactions) {
      context.isRunningReactions = true
      while (context.pending.length) {
        // N.B. errors here cause other pending subscriptions to be aborted!
        context.pending.splice(0).forEach(runFn)
      }
      context.isRunningReactions = false
    }
  }

  function configure(config: Partial<RValConfig>) {
    Object.assign(context.config, config)
  }

  // prettier-ignore
  const api = { val, drv, sub, act, effect, configure }
  return api
}

const defaultPreProcessor = value => value
export const defaultInstance = rval()

class ObservableValue<T> implements ObservableAdministration {
  listeners: Thunk[] = []
  value: T
  get: () => T
  preProcessor: PreProcessor
  constructor(private context: RValContext, public api: RValInstance, state: T, preProcessor) {
    this.get = this.getSet.bind(this) as any
    this.preProcessor = normalizePreProcessor(preProcessor)
    hiddenProp(this.get, $RVal, this)
    this.value = this.freeze(this.preProcessor(state, undefined, this.api))
  }
  addListener(listener) {
    this.listeners.push(listener)
  }
  removeListener(listener) {
    removeCallback(this.listeners, listener)
  }
  getSet(newValue?: T) {
    switch (arguments.length) {
      case 0:
        registerRead(this.context, this)
        return this.value
      case 1:
        // prettier-ignore
        // TODO: re-enable? if (this.context.currentlyComputing) throw new Error('derivations cannot have side effects and update values')
        if(typeof newValue === "function") newValue = newValue(this.value)
        newValue = this.preProcessor(newValue, this.value, this.api) as T
        if (newValue !== this.value) {
          this.value = this.freeze(newValue!)
          this.notifyObservers()
        }
        break
      default:
        throw new Error('val expects 0 or 1 arguments')
    }
  }
  // optimization: could factor out this closure
  notifyObservers = this.api.act(() => {
      runAll(this.listeners)
  })
  freeze(v) {
    if (this.context.config.autoFreeze && (Array.isArray(v) || _isPlainObject(v))) _deepfreeze(v)
    return v
  }
}

class Computed<T = any> implements ObservableAdministration {
  listeners: Thunk[] = []
  inputValues: any[] | undefined = undefined
  observing: Set<ObservableAdministration> = new Set()
  state = NOT_TRACKING
  dirtyCount = 0
  value: T = undefined!
  get: () => T
  setter?: (value) => void
  constructor(private context: RValContext, public api: RValInstance, public derivation: () => T, setter?: (value) => void) {
    this.get = this.getSet.bind(this) as any
    if (setter) this.setter = api.act(setter)
    hiddenProp(this.get, $RVal, this)
  }
  markDirty = () => {
    if (++this.dirtyCount === 1) {
      this.state = STALE
      runAll(this.listeners)
    }
  }
  addListener(observer) {
    this.listeners.push(observer)
  }
  removeListener(observer) {
    removeCallback(this.listeners, observer)
    if (!this.listeners.length) {
      this.observing.forEach(o => o.removeListener(this.markDirty))
      this.observing = new Set()
      this.value = undefined!
      this.state = NOT_TRACKING
      this.inputValues = undefined
    }
  }
  registerDependency(sub: ObservableAdministration) {
    this.observing.add(sub)
  }
  someDependencyHasChanged() {
    switch(this.state) {
      case NOT_TRACKING: return true
      case UP_TO_DATE: return false
      case STALE: 
        if (!inputSetHasChanged(this.observing, this.inputValues)) {
          this.dirtyCount = 0
          this.state = UP_TO_DATE
          return false;
        }
      }
      return true;
  }
  track() {
    if (!this.someDependencyHasChanged()) return
    const oldObserving = this.observing
    const [newValue, newObserving] = track(this.context, this.derivation)
    this.value = newValue
    this.observing = newObserving
    registerDependencies(this.markDirty, oldObserving, newObserving)
    this.inputValues = recordInputSet(newObserving)
    this.dirtyCount = 0
    this.state = UP_TO_DATE
  }
  getSet(value?) {
    switch (arguments.length) {
      case 0:
        // console.log("GET - "+ this.derivation.toString())
        // something being computed? setup tracking
        registerRead(this.context, this)
        // yay, we are up to date!
        if (this.state === UP_TO_DATE) return this.value
        // nope, we are not, and no one is observing either
        if (!this.context.currentlyComputing && !this.listeners.length) {
          // This won't actively remove any listener, but will transition the drv to 
          // untracked, if no other listener arrived
          // TODO: optimize: have one handler for this!
          // TODO: should there be an option to disable this optimization to prevent mem leaking?
          setImmediate(() => this.removeListener(null))
        }
        // maybe scheduled, definitely tracking, value is needed, track now!
        this.track()
        return this.value
      case 1:
        if (this.setter) return void this.setter(value)
    }
    throw new Error("[drv] Didn't expect any arguments");
  }  
}

type DependencySet = (ObservableAdministration | any)[]

function track<R>(context: RValContext, fn: () => R): [R, Set<ObservableAdministration>] {
  const observing = new Set()
  context.currentlyComputingStack.push(observing)
  const res = fn()
  context.currentlyComputingStack.pop()
  return [res, observing]
}

function registerDependencies(listener: Thunk, oldDeps: Set<ObservableAdministration>, newDeps: Set<ObservableAdministration>) {
  // Optimize: 
  if (!oldDeps) {
    newDeps.forEach(d => d.addListener(listener))
  } else {
    newDeps.forEach(o => {
      if (!oldDeps.has(o)) o.addListener(listener)
    })
    oldDeps.forEach(o => {
      if (!newDeps.has(o)) o.removeListener(listener)
    })
  }
}

function registerRead(context: RValContext, observable: ObservableAdministration) {
  // optimize: same last touched by optimization as MobX
  // Sets are used, and keep insertion order, which is important for optimal performance! 
  // (to make someDependencyHasChanged cheap and not
  // re-evaluate deps that might not be needed in the future due some branching logic) 
  if (context.currentlyComputing) context.currentlyComputing.add(observable)
}

function recordInputSet(deps: Set<ObservableAdministration>): any[] {
  // optimize: write more efficiently
  return Array.from(deps).map(currentValue)
}

function inputSetHasChanged(deps: Set<ObservableAdministration>, inputs?: any[]) {
  return !deps || !inputs || !Array.from(deps.values()).every((o, idx) => o.get() === inputs[idx])
}

function currentValue(dep: ObservableAdministration): any {
  // Returns the current, last known (computed) value of a dep
  // Regardless whether that is stale or not 
  return (dep as any).value
}

function runAll(fns: Thunk[]): void {
  fns.forEach(runFn)
}

function runFn(fn: Thunk): void {
  fn()
}

function removeCallback(fns: Thunk[], fn: Thunk) {
  const idx = fns.indexOf(fn)
  if (idx >= 0) fns.splice(idx, 1)
}

function normalizePreProcessor(preProcessor: undefined | PreProcessor | PreProcessor[]): PreProcessor {
  if (!preProcessor) return defaultPreProcessor
  if (typeof preProcessor === "function") return preProcessor
  if (Array.isArray(preProcessor))
    return function(newValue, currentValue, api) {
      return preProcessor.reduce((acc, current) => current(acc, currentValue, api), newValue)
    }
  throw new Error("No valid preprocessor");
}

export function isVal(value: any): value is Val {
  return typeof value === "function" && value[$RVal] instanceof ObservableValue
}

export function isDrv(value: any): value is Drv {
  return typeof value === "function" && value[$RVal] instanceof Computed
}

export function _once<T extends Function>(fn: T): T {
  // based on 'once' package, but made smaller
  var f: any = function(this: any) {
    if (f.called) return f.value
    f.called = true
    return (f.value = fn.apply(this, arguments))
  }
  f.called = false
  return f
}

export function _deepfreeze(o) {
  // based on 'deepfreeze' package, but copied here to simplify build setup :-/
  if (o === Object(o)) {
    Object.isFrozen(o) || Object.freeze(o)
    Object.getOwnPropertyNames(o).forEach(function(prop) {
      prop === 'constructor' || _deepfreeze(o[prop])
    })
  }
  return o
}

function hiddenProp(target, key, value) {
  Object.defineProperty(target, key, {
    configurable: true,
    enumerable: false,
    writable: false,
    value: value
  })
}

export function _isPlainObject(o) {
  const p = o && typeof o === "object" && Object.getPrototypeOf(o)
  return p === Object.prototype || p === null
}

export const val = defaultInstance.val
export const drv = defaultInstance.drv
export const sub = defaultInstance.sub
export const act = defaultInstance.act
export const effect = defaultInstance.effect
export const configure = defaultInstance.configure
