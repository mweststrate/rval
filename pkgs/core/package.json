{
  "name": "@r-val/core",
  "private": false,
  "version": "0.0.4",
  "description": "minimalistic, transparent reactive programming library",
  "main": "dist/core.js",
  "umd:main": "dist/core.umd.js",
  "unpkg": "dist/core.umd.js",
  "module": "dist/core.mjs",
  "jsnext:main": "dist/core.mjs",
  "react-native": "dist/core.mjs",
  "browser": {
    "./dist/core.js": "./dist/core.js",
    "./dist/core.mjs": "./dist/core.mjs"
  },
  "source": "core.ts",
  "scripts": {},
  "author": "Michel Weststrate",
  "license": "MIT",
  "files": [
    "*.ts",
    "dist"
  ],
  "reserved": [
    "configurable",
    "enumerable",
    "writable",
    "value",
    "$RVal",
    "rval",
    "defaultInstance",
    "isVal",
    "isDrv",
    "_deepfreeze",
    "_once",
    "_isPlainObject",
    "val",
    "drv",
    "sub",
    "run",
    "act",
    "effect",
    "configure",
    "autoFreeze",
    "fireImmediately"
  ],
  "mangle": {
    "reserved": [
      "configurable",
      "enumerable",
      "writable",
      "value",
      "$RVal",
      "rval",
      "defaultInstance",
      "isVal",
      "isDrv",
      "_deepfreeze",
      "_once",
      "_isPlainObject",
      "val",
      "drv",
      "sub",
      "run",
      "act",
      "effect",
      "configure",
      "autoFreeze",
      "fireImmediately"
    ]
  },
  "gitHead": "a824fe9bd83228fb126c8c6188ef7b85fc2ffdd7"
}