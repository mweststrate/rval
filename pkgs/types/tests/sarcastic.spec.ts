import { val } from "@r-val/core"

import * as is from "sarcastic"


test('isNumber', () => {
    const v = val<any>(3, is.number)
  
    expect(() => v("4")).toThrow("3 must be a number")
    expect(v()).toBe(3)
    v(4)
    expect(v()).toBe(4)
})
  

describe('isShape', () => {
    const todoShape = is.shape({
      done: is.boolean,
      title: is.string
    })
  
    const todoStoreShape = is.shape({
      todos: is.arrayOf(todoShape)
    })
  
    test('simple', () => {
      const t1 = val({ done: false, title: "get coffee"}, todoShape)
      expect(() => t1({} as any)).toThrowErrorMatchingSnapshot()
      expect(() => t1({ done: true, title: 3 } as any)).toThrowErrorMatchingSnapshot()
      t1({
        extra: 7,
        done: false,
        title: "stuff"
      } as any)
    })
  
    test('complex', () => {
      const store = val({
        todos: [{ done: false, title: "get coffee"}]
      }, todoStoreShape)
  
      expect(() => store({} as any)).toThrowErrorMatchingSnapshot()
      store({
        todos: []
      })
      expect(() => store({
        todos: [{ done: false, title: "get coffee"}, 3]
      } as any)).toThrowErrorMatchingSnapshot()
    })
  })