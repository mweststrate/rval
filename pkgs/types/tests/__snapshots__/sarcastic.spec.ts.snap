// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isShape complex 1`] = `"[object Object].todos must be an array"`;

exports[`isShape complex 2`] = `"[object Object].todos[1] must be an object"`;

exports[`isShape simple 1`] = `"[object Object].done must be a boolean"`;

exports[`isShape simple 2`] = `"[object Object].title must be a string"`;
